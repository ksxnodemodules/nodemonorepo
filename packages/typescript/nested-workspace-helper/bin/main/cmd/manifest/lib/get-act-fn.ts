import ramda from 'ramda'

export type ActionChoice = 'set' | 'delete' | 'assign'
export type ActionFunction = (manifest: object, path: string[], value?: any) => object

export function getActionFunction (action: ActionChoice): ActionFunction {
  switch (action) {
    case 'set':
      return (object, path, value) =>
        ramda.assocPath(path, value, object)
    case 'delete':
      return (object, path) =>
        ramda.dissocPath(path, object)
    case 'assign':
      return (object, path, value) =>
        ramda.assocPath(path, {...ramda.path(path) || {}, ...value}, object)
  }
}

export default getActionFunction
