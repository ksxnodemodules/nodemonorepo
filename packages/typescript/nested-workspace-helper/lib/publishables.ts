import ramda from 'ramda'
import * as assets from 'monorepo-shared-assets'
import createRegistryFactory from './npm-registry'
import listAllPackages from './list-pkgs'

import {
  Package,
  PackageRegistryResponse,
  PublishableClassification
} from './types'

export type PartitionResult = Promise<PublishableClassification>

/**
 * Classify packages in directory into 3 groups
 *   - `publishables`: Public packages that can be published (version does not exist in registry)
 *   - `unpublishables`: Public packages that are not publishable (version exists in registry)
 *   - `skip`: Private and/or anonymous packages that are not meant to be published
 * @param dirname Directory that contains all packages
 * @param registry URL origin of registry (e.g. `http://registry.npmjs.org`)
 * @returns The three groups of packages
 */
export async function partition (dirname: string, registry?: string) {
  return partition.fromPackageList(await listAllPackages(dirname), registry)
}

export namespace partition {
  /**
   * Classify packages from list
   * @param list List of packages
   * @param registry URL origin of registry (e.g. `http://registry.npmjs.org`)
   * @returns The three groups of packages
   * @see partition
   */
  export async function fromPackageList (list: Package.List, registry?: string): PartitionResult {
    enum Classification {
      publishable,
      unpublishable,
      skip
    }

    const {getAllVersions} = createRegistryFactory(registry)

    const labeled = await Promise.all(
      list.map(async pkg => {
        const {private: prv, name = '', version} = pkg.manifestContent
        const res = (cls: Classification) => ({pkg, cls})

        if (prv || !name) return res(Classification.skip)
        if (!version) return res(Classification.unpublishable)

        const published = await getAllVersions(name)
        if (published === 'NotFound') return res(Classification.publishable)
        if (published.versions[version]) return res(Classification.unpublishable)

        return res(Classification.publishable)
      })
    )

    const classification = assets.group.classify.dict.singleDistribute(labeled, x => String(x.cls))

    const getpkgs = (id: Classification) => {
      const list = classification[id]
      return list ? list.map(x => x.pkg) : []
    }

    return {
      publishables: getpkgs(Classification.publishable),
      unpublishables: getpkgs(Classification.unpublishable),
      skip: getpkgs(Classification.skip)
    }
  }

  /**
   * Create a factory base on registry
   * @param registry URL origin of registry (e.g. `http://registry.npmjs.org`)
   * @returns A collection of functions
   */
  export function createFactory (registry?: string) {
    const param: [typeof registry] = [registry]

    return Object.assign(
      ramda.partialRight(partition, param),
      {
        fromPackageList: ramda.partialRight(fromPackageList, param)
      }
    )
  }
}

export default partition
