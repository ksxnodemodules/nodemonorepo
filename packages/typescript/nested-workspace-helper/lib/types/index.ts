/**
 * Basic types that other types base upon
 */
export namespace Basic {
  export type Path = string
  export type PackageName = string
  export type PackageVersion = string
  export type PackageVersionRequirement = string
  export type ModulePath = string
}

/**
 * Functions that take `X` and return `Y`
 */
export type Transformer<X, Y> = (x: X) => Y

export namespace Transformer {
  /**
   * Functions that take `Package.ListItem` and return `Y`
   */
  export type FromListItem<Y> = Transformer<Package.ListItem, Y>

  export namespace FromListItem {
    /**
     * Functions that take `Package.ListItem` and return `Package.Manifest`
     */
    export type ToManifest = FromListItem<Package.Manifest>
  }
}

/**
 * Basic types for packages
 */
export namespace Package {
  /**
   * Structure found in `dependencies`, `devDependencies` and `peerDependencies` in `package.json` files
   */
  export interface Dict {
    readonly [name: string]: Basic.PackageVersionRequirement
  }

  /**
   * Structure of `package.json`
   */
  export interface Manifest {
    readonly name?: Basic.PackageName
    readonly version?: Basic.PackageVersion
    readonly private?: boolean
    readonly main?: Basic.ModulePath
    readonly dependencies?: Dict
    readonly devDependencies?: Dict
    readonly peerDependencies?: Dict
    readonly [_: string]: any
  }

  export namespace Manifest {
    export namespace Writer {
      /**
       * Options to be passed to `writePackageManifest`/`writePackageManifest.write`
       */
      export interface Options {
        /**
         * Indentation of `package.json`
         * - As number, it is number of spaces per indentation unit
         * - As string, it is literally indentation unit
         */
        readonly indentation?: string | number

        /**
         * How many final new line to be inserted? Default is 1
         */
        readonly finalNewLine?: number
      }
    }
  }

  /**
   * Returning value of `listAllPackages`
   * @see listAllPackages
   */
  export type List = ReadonlyArray<ListItem>

  export interface ListItem {
    readonly path: Basic.Path
    readonly manifestFile: Basic.Path
    readonly manifestContent: Manifest
  }
}

/**
 * Representation of a dependency listed in `package.json`
 */
export interface Dependency {
  /** Name of dependency, it is property name in dependant's `package.json` dependencies object */
  readonly name: Dependency.Name

  /** Dependency's actual version, it is field `version` in dependency's own `package.json` */
  readonly version: Dependency.Version

  /** Whether dependency is `prod`, `dev`, or `peer` */
  readonly type: Dependency.Type

  /** Requirement of dependency, it is property value in dependant's `package.json` dependencies object */
  readonly requirement: Dependency.Requirement

  /** Other information */
  readonly info: Package.ListItem
}

export namespace Dependency {
  export type Name = Basic.PackageName
  export type Version = Basic.PackageVersion
  export type Type = 'prod' | 'dev' | 'peer'
  export type Requirement = Basic.PackageVersionRequirement

  export type List = ReadonlyArray<Dependency>

  /**
   * Dependant-dependency map within a project (directory)
   *   - Key: Directory that contain dependant's `package.json`
   *   - Value:
   *     - `list`: List of dependencies
   *     - `dependant`: Dependant's information
   */
  export interface Map {
    readonly [dirname: string]: MapValue
  }

  export interface MapValue {
    readonly list: List
    readonly dependant: Package.ListItem
  }
}

export namespace MismatchedDependency {
  export interface ListItem extends Dependency {
    update: Basic.PackageVersionRequirement
  }

  export type List = ReadonlyArray<ListItem>

  export interface MapValue {
    readonly list: List,
    readonly dependant: Package.ListItem
  }

  /**
   * Map of mismatched dependency
   * Structure similar to `DependencyMap`
   * @see DependencyMap
   */
  export interface Map {
    readonly [name: string]: MapValue
  }
}

/**
 * Status when interact in network (e.g. Internet)
 */
export type NetworkStatus = NetworkStatus.NotFound

/**
 * Collection of network statuses
 */
export namespace NetworkStatus {
  /** When status code is `404` */
  export type NotFound = 'NotFound'
}

/**
 * Common properties of registry response object
 */
export interface Registry {
  /** Package name */
  readonly name: string

  /** Package description */
  readonly description?: string

  /** Other properties */
  readonly [_: string]: any
}

/**
 * Registry information of a package
 * including all of its versions
 */
export interface PackageRegistry extends Registry {
  readonly versions: RegistryPackageVersionSet
}

export type PackageRegistryResponse = PackageRegistry | NetworkStatus

/**
 * Registry information of a package of a single version
 */
export interface PackageVersionRegistry extends RegistryPackageManifest {}

export type PackageVersionRegistryResponse = PackageVersionRegistry | NetworkStatus

/**
 * All versions of a single package in registry
 */
export interface RegistryPackageVersionSet {
  readonly [version: string]: PackageVersionRegistry
}

/**
 * All packages in registry
 */
export interface RegistryPackageSet {
  readonly [name: string]: PackageRegistry
}

export interface RegistryPackageManifest extends Package.Manifest {
  readonly version: string
}

/** Contains group of publishable, unpublishable and private packages */
export interface PublishableClassification {
  /** Public packages that can be publishes */
  readonly publishables: PublishablePackageList

  /** Public packages that have yet to be publishable */
  readonly unpublishables: UnpublisablePackageList

  /** Private and/or anonymous packages */
  readonly skip: PrivatePackageList
}

export type PublishablePackageList = Package.List
export type UnpublisablePackageList = Package.List
export type PrivatePackageList = Package.List

export namespace InvalidPackage {
  export type List = ReadonlyArray<ListItem>

  export interface ListItem<ReasonElement = Reason> extends Package.ListItem {
    readonly reason: ReadonlyArray<ReasonElement>
  }

  export type Reason =
    Reason.InvalidDependencies |
    Reason.PrivateDependencies |
    Reason.NameDuplication |
    Reason.SelfDependence

  export namespace Reason {
    export abstract class Base {
      /**
       * Information accessible from classes themselve
       */
      static readonly description: Base.Description

      /**
       * Name of reason
       *   - Is the name of class when not compressed
       *   - More reliant than class names
       */
      readonly name: string

      /**
       * Helpful description that can be use as error message
       */
      readonly message: string

      get info () {
        return (this.constructor as typeof Base).description
      }

      constructor () {
        const {description} = this.constructor as typeof Base
        this.name = description.name
        this.message = description.message
      }
    }

    export namespace Base {
      export class Description {
        readonly name: string
        readonly message: string

        constructor (name: string, message: string) {
          this.name = name
          this.message = message
        }
      }
    }

    export class InvalidDependencies extends Base {
      static readonly description = new Base.Description(
        'InvalidDependencies',
        'Package depends on invalid dependencies'
      )

      /**
       * List of invalid dependecies that the package depended on
       */
      readonly dependencies: InvalidDependencies.DependencyList

      /**
       * @param dependencies Intended value of `this.dependencies`
       */
      constructor (dependencies: InvalidDependencies.DependencyList) {
        super()
        this.dependencies = dependencies
      }
    }

    export namespace InvalidDependencies {
      export type DependencyList = ReadonlyArray<ListItem>
    }

    export class PrivateDependencies extends Base {
      static readonly description = new Base.Description(
        'PrivateDependencies',
        'Public package depends on non-dev private dependencies'
      )

      /**
       * List of private dependencies that this package depended on as prod or peer
       */
      readonly dependencies: PrivateDependencies.DependencyList

      /**
       * @param dependencies Intended value of `this.dependencies`
       */
      constructor (dependencies: PrivateDependencies.DependencyList) {
        super()
        this.dependencies = dependencies
      }
    }

    export namespace PrivateDependencies {
      export type DependencyList = ReadonlyArray<Dependency>
    }

    export class NameDuplication extends Base {
      static readonly description = new Base.Description(
        'NameDuplication',
        'Package with name that is used by another package'
      )

      /**
       * List of packages that used the name
       */
      readonly duplicates: NameDuplication.PackageList

      constructor (duplicates: NameDuplication.PackageList) {
        super()
        this.duplicates = duplicates
      }
    }

    export namespace NameDuplication {
      export type PackageList = ReadonlyArray<Package.ListItem>
    }

    export class SelfDependence extends Base {
      static readonly description = new Base.Description(
        'SelfDependence',
        'Package that depends on itself'
      )
    }
  }
}
