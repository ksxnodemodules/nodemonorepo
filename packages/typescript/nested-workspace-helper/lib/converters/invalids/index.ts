import chalk from 'chalk'
import * as assets from 'monorepo-shared-assets'
import {InvalidPackage} from '../../types'
import * as pkgUtils from '../../utils/pkg'

export function group (list: InvalidPackage.List) {
  return group.asMap(list)
}

export namespace group {
  export function asDict (list: InvalidPackage.List) {
    return assets.group.classify.dict.multiDistribute(
      list,
      item => item.reason.map(x => x.name),
      utils.DUPLICATION_CHECKER
    ).classified
  }

  export namespace asDict {
    export function simple (list: InvalidPackage.List): simple.Dict {
      const result: {
        [name: string]: simple.Dict.Value
      } = {}

      for (const [name, pkgs] of Object.entries(asDict(list))) {
        if (!pkgs) continue

        result[name] = pkgs.map(x => ({
          name: x.manifestContent.name,
          path: x.path,
          causes: Array.from(new Set(
            x.reason
              .map(asString.getReasonCauses)
              .reduce((prev, current) => [...prev, ...current])
          ))
        }))
      }

      return result
    }

    export namespace simple {
      export interface Dict {
        readonly [name: string]: Dict.Value
      }

      export namespace Dict {
        export type Value = ReadonlyArray<Value.Element>

        export namespace Value {
          export interface Element {
            readonly name?: string
            readonly path: string
            readonly causes: ReadonlyArray<string>
          }
        }
      }
    }
  }

  export function asMap (list: InvalidPackage.List) {
    return assets.group.classify.map.multiDistribute(
      list,
      item => item.reason.map(
        x => x.constructor as typeof InvalidPackage.Reason.Base
      ),
      undefined,
      utils.DUPLICATION_CHECKER
    ).classified
  }

  export function asString (list: InvalidPackage.List): string {
    return asString.fromClassifiedMap(group(list))
  }

  export namespace asString {
    export function fromClassifiedMap (map: ClassifiedMap): string {
      return Array.from(map).map(([cls, pkgs]) => {
        const {name, message} = cls.description
        const heading = indent.lv[0] + message + chalk.bold(` (${name})`)

        const content = Array.from(pkgs).map(item => {
          const heading = indent.lv[1] + pkgUtils.name(item) + chalk.dim(` (${item.path})`)

          const content = item.reason.map(
            reason => asString
              .getReasonCauses(reason)
              .filter(x => x.length)
              .map(x => indent.lv[2] + x)
              .join('\n')
          ).filter(x => x.trim().length)

          return utils.heading.concat(heading, Array.from(new Set(content)))
        })

        return utils.heading.concat(heading, content)
      }).join('\n\n')
    }

    export function getReasonCauses (reason: InvalidPackage.Reason): string[] {
      const getStringArray = () => {
        const {
          InvalidDependencies,
          NameDuplication,
          PrivateDependencies,
          SelfDependence
        } = InvalidPackage.Reason

        if (reason instanceof InvalidDependencies) {
          return reason.dependencies.map(x => x.manifestContent.name || '[anonymous]')
        }

        if (reason instanceof NameDuplication) {
          return reason.duplicates.map(x => x.path)
        }

        if (reason instanceof PrivateDependencies) {
          return reason.dependencies.map(x => x.name)
        }

        if (reason instanceof SelfDependence) {
          return []
        }

        throw new Error('Invalid type of reason')
      }

      return Array.from(new Set(getStringArray()))
    }
  }

  export type ClassifiedMap = ReadonlyMap<
    typeof InvalidPackage.Reason.Base,
    ReadonlySet<InvalidPackage.ListItem>
  >

  namespace indent {
    const character = ' '
    const prefixes = '*—→'

    export const lv = Array
      .from(assets.iter.fns.range(3))
      .map(x => character.repeat(x << 1) + prefixes[x] + ' ')
  }
}

namespace utils {
  export namespace heading {
    export function concat (
      heading: string,
      content: ReadonlyArray<string>
    ): string {
      return content.length
        ? heading + '\n' + content.join('\n')
        : heading
    }
  }

  export function DUPLICATION_CHECKER (
    a: InvalidPackage.ListItem,
    b: InvalidPackage.ListItem
  ): boolean {
    return a.path === b.path
  }
}
