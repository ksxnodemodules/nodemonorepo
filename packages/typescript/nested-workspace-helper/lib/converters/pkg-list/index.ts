import chalk from 'chalk'
import * as types from '../../types'
import * as pkgUtils from '../../utils/pkg'

export namespace list {
  export function asString (list: types.Package.List): string {
    return asStringArray(list).join('\n')
  }

  export function asStringArray (list: types.Package.List): string[] {
    return list.map(x => chalk.bold(pkgUtils.name(x)) + chalk.dim(` (${x.path})`))
  }
}
