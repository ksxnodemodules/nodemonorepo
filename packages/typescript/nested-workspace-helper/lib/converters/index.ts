import * as pkgList from './pkg-list'
import * as invalids from './invalids'
export {pkgList, invalids}
