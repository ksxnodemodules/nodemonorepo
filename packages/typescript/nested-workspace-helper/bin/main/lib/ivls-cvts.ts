import * as converters from '../../../lib/converters/invalids'
export const group = converters.group
export const getMap = group.asMap
export const getDict = group.asDict
export const getString = group.asString
