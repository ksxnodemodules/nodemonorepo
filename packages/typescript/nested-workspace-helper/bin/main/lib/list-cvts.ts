import {list} from '../../../lib/converters/pkg-list'
export const getString = list.asString
