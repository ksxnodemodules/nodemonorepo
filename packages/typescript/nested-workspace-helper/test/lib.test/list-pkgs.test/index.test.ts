import {listAllPackages} from '../../../index'
import * as converters from '../../../lib/converters'
import createSetupTeardown from '../../.lib/setup-teardown'

const {apply} = createSetupTeardown('valid.yaml')
const getPromise = () => listAllPackages('root')

describe('listAllPackages function', () => {
  it('matches snapshot', apply(async () => {
    expect(await getPromise()).toMatchSnapshot()
  }))
})

describe('coverter function', () => {
  it('convert list to string', apply(async () => {
    expect(converters.pkgList.list.asString(await getPromise())).toMatchSnapshot()
  }))

  it('convert list to arrays', apply(async () => {
    expect(converters.pkgList.list.asStringArray(await getPromise())).toMatchSnapshot()
  }))
})
