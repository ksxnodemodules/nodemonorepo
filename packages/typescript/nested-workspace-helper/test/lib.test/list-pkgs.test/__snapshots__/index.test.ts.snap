// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`coverter function convert list to arrays 1`] = `
Array [
  "[1mroot[22m[2m (root)[22m",
  "[1mpkg00[22m[2m (root/wrkspc0dept/pkg00)[22m",
  "[1mpkg01[22m[2m (root/wrkspc0dept/pkg01)[22m",
  "[1mpkg100[22m[2m (root/wrkspc1dept/sub10/pkg100)[22m",
  "[1mpkg110[22m[2m (root/wrkspc1dept/sub11/pkg110)[22m",
  "[1mpkg111prv[22m[2m (root/wrkspc1dept/sub11/pkg111prv)[22m",
  "[1mpkg2000[22m[2m (root/wrkspc2dept/sub20/sub200/pkg2000)[22m",
]
`;

exports[`coverter function convert list to string 1`] = `
"[1mroot[22m[2m (root)[22m
[1mpkg00[22m[2m (root/wrkspc0dept/pkg00)[22m
[1mpkg01[22m[2m (root/wrkspc0dept/pkg01)[22m
[1mpkg100[22m[2m (root/wrkspc1dept/sub10/pkg100)[22m
[1mpkg110[22m[2m (root/wrkspc1dept/sub11/pkg110)[22m
[1mpkg111prv[22m[2m (root/wrkspc1dept/sub11/pkg111prv)[22m
[1mpkg2000[22m[2m (root/wrkspc2dept/sub20/sub200/pkg2000)[22m"
`;

exports[`listAllPackages function matches snapshot 1`] = `
Array [
  Object {
    "manifestContent": Object {
      "devDependencies": Object {
        "pkg00": "^0.1.0",
      },
      "name": "root",
      "private": true,
      "version": "0.0.0",
      "workspaces": Array [
        "wrkspc0dept/*",
        "wrkspc1dept/*/*",
        "wrkspc2dept/*/*/*",
      ],
    },
    "manifestFile": "root/package.json",
    "path": "root",
  },
  Object {
    "manifestContent": Object {
      "name": "pkg00",
      "version": "0.1.2",
    },
    "manifestFile": "root/wrkspc0dept/pkg00/package.json",
    "path": "root/wrkspc0dept/pkg00",
  },
  Object {
    "manifestContent": Object {
      "name": "pkg01",
      "private": false,
      "version": "1.2.3",
    },
    "manifestFile": "root/wrkspc0dept/pkg01/package.json",
    "path": "root/wrkspc0dept/pkg01",
  },
  Object {
    "manifestContent": Object {
      "name": "pkg100",
      "version": "2.2.5",
    },
    "manifestFile": "root/wrkspc1dept/sub10/pkg100/package.json",
    "path": "root/wrkspc1dept/sub10/pkg100",
  },
  Object {
    "manifestContent": Object {
      "dependencies": Object {
        "pkg01": "^1.0.0",
        "pkg100": "~2.2.1",
      },
      "name": "pkg110",
      "version": "22.3.5",
    },
    "manifestFile": "root/wrkspc1dept/sub11/pkg110/package.json",
    "path": "root/wrkspc1dept/sub11/pkg110",
  },
  Object {
    "manifestContent": Object {
      "dependencies": Object {
        "pkg01": "*",
      },
      "name": "pkg111prv",
      "private": true,
      "version": "33.22.11",
    },
    "manifestFile": "root/wrkspc1dept/sub11/pkg111prv/package.json",
    "path": "root/wrkspc1dept/sub11/pkg111prv",
  },
  Object {
    "manifestContent": Object {
      "devDependencies": Object {
        "foreign-foo": "^1.2.3",
        "pkg111prv": "33.22.11",
      },
      "name": "pkg2000",
      "version": "123.456.789",
    },
    "manifestFile": "root/wrkspc2dept/sub20/sub200/pkg2000/package.json",
    "path": "root/wrkspc2dept/sub20/sub200/pkg2000",
  },
]
`;
